recipients-resolver/src/main/java/com/redhat/cloud/notifications/recipients/resolver/RecipientsResolver.java

```diff
@@ -1,93 +1,92 @@
+        users.removeIf(user -> {
 
-        // If the user preferences should be ignored, the recipients from RBAC (possibly filtered by the
-        // previous step) is returned without filtering out the users who didn't subscribe to the event type.
-        if (request.isIgnoreUserPreferences()) {
-            users = Set.copyOf(users);
-        } else {
-            if (!subscribedByDefault) {
-                // Otherwise, the recipients from RBAC who didn't subscribe to the event type are filtered out.
-                users = filterUsers(users, subscribers);
-            } else {
-                users = filterUnsubscribedUsers(users, subscribers);
+            /*
+             * When the request contains a list of users, only these users will qualify as recipients,
+             * if we did fetch them from the external service. Any fetched users who are not included
+             * in the request are removed.
+             */
+            if (!request.getUsers().isEmpty() && !containsIgnoreCase(request.getUsers(), user.getUsername())) {
+                return true;
             }
-        }
 
-        return users;
-    }
+            // Subscriptions are only considered if the user preferences are NOT ignored.
+            if (!request.isIgnoreUserPreferences()) {
+                if (subscribedByDefault) {
+                    // When subscribedByDefault is true, we need to remove from the users anyone who unsubscribed.
+                    return containsIgnoreCase(unsubscribers, user.getUsername());
+                } else {
+                    // When subscribedByDefault is false, we need to remove from the users anyone who DIT NOT subscribe.
+                    return !containsIgnoreCase(subscribers, user.getUsername());
+                }
+            }
 
-    private Set<User> filterUsers(Set<User> users, Set<String> target) {
-        return users.stream()
-            .filter(
-                user -> target
-                    .stream()
-                    .anyMatch(requested -> requested.equalsIgnoreCase(user.getUsername()))
-            )
-            .collect(Collectors.toSet());
-    }
+            // The current user was not removed by any of the conditions above, so we'll keep them in the Set!
+            return false;
 
-    private Set<User> filterUnsubscribedUsers(Set<User> users, Set<String> target) {
-        return users.stream()
-            .filter(
-                user -> target
-                    .stream()
-                    .noneMatch(requested -> requested.equalsIgnoreCase(user.getUsername()))
-            )
-            .collect(Collectors.toSet());
+        });
+
+        return users;
     }
 
+    private static boolean containsIgnoreCase(Set<String> usernames, String username) {
+        if (usernames == null || usernames.isEmpty()) {
+            return false;
+        } else {
+            return usernames.stream()
```

Comment:
Couldn't we transform subscribers and unsubscribers only once at the beginning of calling method ?
Thanks to that, we may also initialise once an empty Set if subscribers or unsubscribers is null

Reply:
Good point. I pushed something.